flocker/node/test/test_loop.py

```diff
@@ -319,15 +321,57 @@ def test_convergence_done_notify(self):
         self.assertEqual(client.calls, [(NodeStateCommand,
                                          dict(node_state=local_state))])
 
+    def test_convergence_done_update_local_state(self):
+        """
+        An FSM doing convergence that gets a discovery result updates the local
+        representation of the cluster state before calculating necessary state
+        changes.
+        """
+        local_node_hostname = u'192.0.2.123'
+        # Control service reports that this node has no manifestations.
+        received_node = Node(hostname=local_node_hostname)
+        received_cluster_state = Deployment(
+            nodes=frozenset([received_node])
+        )
+        discovered_manifestation = Manifestation(
+            dataset=Dataset(dataset_id=uuid4()),
+            primary=True
+        )
+        local_node_state = NodeState(
+            hostname=local_node_hostname,
+            manifestations=[discovered_manifestation]
+        )
+        client = self.successful_amp_client([local_node_state])
+        action = ControllableAction(Deferred())
+        deployer = ControllableDeployer([succeed(local_node_state)], [action])
+
+        loops = []
+        def spy_loop_factory(reactor, deployer):
```

Comment:
Instead of testing this way, I suggest passing in a `ControllableDeployer` and making sure its calculate function is called with updated state... both in this loop *and* in a future loop. The principle here is that as far as possible one should test public interfaces, and the interface in this case is that the IDeployer calculate function is called.

Additionally, `ControllableDeployer` should have its `IDeployer`-ness verified using `ideployer_tests_factory` from `test_deploy.py`.